### PR TITLE
AnimatedVisualPlayer destructor bug

### PR DESCRIPTION
**Pull request by AndrewB330**
This change should fix the [internal bug](https://microsoft.visualstudio.com/OS/_workitems/edit/32126845/)

The problem is that calling `IsPlaying(false)` inside the `AnimatedVisualPlayer`s destructor causes a crash. Basically, our goal is to reduce a number of things that destructor is doing, ideally nothing. But the problem is that we want to destroy `AnimationPlay m_nowPlaying` object (if it is alive) as well, but `AnimatedVisualPlayer::PlayAsync` keeps this object alive, because it holds reference and awaits (`co_await *thisPlay;`) for it to call `CompleteAwaits()`. Also we need to release `m_batch` listener as well, because it holds weak reference to us (line ~130).

So basically, we want to call `AnimationPlay ::Complete()` but without touching `m_owner`. But since `AnimationPlay` calls `m_owner` only if it is bound to it, we can detach it first, to ensure that we will not call anything in m_owner inside the destructor.

### CODE REVIEW

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp [hunk 1]
```diff
@@ -342,10 +343,14 @@ AnimatedVisualPlayer::AnimatedVisualPlayer()
...
-    // NOTE: Stop() can cause reentrance when clients react to DP changes, but because
-    //       we're in the destructor we know that there aren't any clients who can reach
-    //       us, so reentrance is not a concern. 
-    Stop();
+    if (m_nowPlaying)
+    {
+        // To stop and destroy m_nowPlaying we need to call Complete()
+        // We need to detach m_nowPlaying first so that Complete()
+        // will not try to call us (owner), since we already in the destructor
+        auto nowPlaying = std::move(m_nowPlaying);
```
Comment (jevansaks) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:351]: How does the move "detach" it? m_nowPlaying still points back at AnimatedVisualPlayer, right?

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp [hunk 2]
```diff
@@ -342,10 +343,14 @@ AnimatedVisualPlayer::AnimatedVisualPlayer()
...
+        nowPlaying->Complete();
+    }
```
Comment (jevansaks) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:353]: Can whatever cleanup needs to happen just be in the destructor of this object?

Comment (AndrewB330) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:353]: As I see `AnimatedVisualPlayer::PlayAsync` [(code)](https://github.com/microsoft/microsoft-ui-xaml/pull/4955/commits/a2dadc322c90ea5a7b46bbc923357f051fed5ba8#L642) holds reference to this object too, and it will not release it until we call `CompleteAwaits()`, because of `co_await *thisPlay` [(code)](https://github.com/microsoft/microsoft-ui-xaml/pull/4955/commits/a2dadc322c90ea5a7b46bbc923357f051fed5ba8#L660), is it right?

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp (hunk 1 above)
Comment (AndrewB330) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:351]: std::move leaves std::shared_ptr in empty state (nullptr)

Comment (scott-moore-ms) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:351]: After the move, m_nowPlaying is null
Thus when the helper checks IsCurrentPlay(), that will return false and avoid calling back into the AVP to update its state.

Comment (jevansaks) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:351]: But it still calls m_owner to do so. So if NowPlaying outlives AnimatedVisualPlayer, we're still in trouble.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp (hunk 2 above)
Comment (jevansaks) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:353]: If we think that AnimationPlay outlives AnimatedVisualPlayer then AnimationPlay should reference AnimatedVisualPlayer with winrt::weak_ref and check it when it uses it. 

The comment should also be updated to be more precise -- it won't call the owner but it will check the owner's m_nowPlaying field.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp (hunk 1 above)
Comment (AndrewB330) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:351]: `nowPlaying->Complete()` is called inside this destructor, so [this check](https://github.com/microsoft/microsoft-ui-xaml/pull/4955/commits/a2dadc322c90ea5a7b46bbc923357f051fed5ba8#R144) will happen before destructor actually destroys the object?

Comment (jevansaks) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:351]: Yes but it's an accident waiting to happen. It might be that the code today never touches m_owner after this point but what if it does? Can we instead make the code clear the m_owner field out of m_nowPlaying?

```
m_nowPlaying->m_owner = nullptr;
// Call complete after clearing m_owner so we don't get called back to update IsPlaying
m_nowPlaying->Complete();
```

Then when this object finishes destructing it'll release its shared_ptr ref on m_nowPlaying and it's obvious from the code that we are preventing a dangling pointer situtaion.

Comment (AndrewB330) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:351]: I see..., yeah it makes sense

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp (hunk 2 above)
Comment (AndrewB330) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:353]: I think that since AnimationPlay is declared inside AnimatedVisualPlayer, I can't use weak_ref here?

`an undefined class is not allowed as an argument to compiler intrinsic type trait '__is_base_of'`

Comment (jevansaks) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp:353]: My other suggestion of nulling m_owner in AnimatedVisualPlayer destructor is sufficient. Then the pointer won't be dangling after destruction.

**Review by jevansaks — APPROVED** [review]
